# Map, Filter, Reduce, Lambda & Recursion

## Tasks Today:

1) <b>Lambda Functions</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Saving to a Variable <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Multiple Inputs <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Passing a Lambda into a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; e) Returning a Lambda from a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; f) In-Class Exercise #1 <br>
2) <b>Map</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Map <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #2 <br>
3) <b>Filter</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Filter <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #3 <br>
4) <b>Reduce</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Reduce <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #4 <br>
5) <b>Recursion</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Implementing a Base <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Writing a Factorial Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #5 <br>
6) <b>Generators & Iterators</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Yield Keyword <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Inifinite Generator <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #6 <br>
7) <b>Exercises</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Exercise #1 - Filtering Empty Strings <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Exercise #2 - Sorting with Last Name <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Exercise #3 - Conversion to Farhenheit <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Exercise #4 - Fibonacci Sequence <br>

## Lambda Functions <br>
<p>Lambda functions... or "Anonymous Functions" are referring to inline functions with no name. The keyword lambda denotes the no name function, and executes within a single line. Without saving it to a variable; however, it is not able to be used, unless passed in either as a paramater or within list comprehension.<br>Written as "(keyword lambda) (one or more inputs) (colon) (function to be executed)"</p>

#### Syntax

In [1]:
lambda num : num + 2
(lambda num : num + 2)(4)

6

In [ ]:
def add_two(num):
    return num + 2

#### Saving to a Variable

In [2]:
add_two_lambda = lambda num : num + 2

add_two_lambda(10)

12

#### Multiple Inputs

In [3]:
(lambda num1, num2, num3 : num1 + num2 + num3)(6,4,10)

20

In [5]:
add_three_numbers = lambda num1, num2, num3 : num1 + num2 + num3

add_three_numbers(6,4,10)

20

In [ ]:
def add_three_func(num1, num2, num3):
    return num1 + num2 + num3

#### Passing a Lambda into a Function

In [16]:
def math_operations(f, num):
    return f(num)

math_operations(lambda x : x**2 , 10)

math_operations(lambda num: num//10, 1000)

multiply_by_ten = lambda num : num * 10
math_operations(multiply_by_ten, 5)

50

#### Returning a Lambda from a Function

In [13]:
def outter_func():
    outter_num = 4
    def multiply(num):
        inner_num = 10
        return num * outter_num
    return multiply
    
print(outter_func()(4))

my_multiply = outter_func()

my_multiply(4)

def return_lambda(num1, num2):
    return lambda num3, num4 : num1 + num2 + num3 + num4

add_two_three = return_lambda(2,3)

add_two_three(10,20)

16


35

In [33]:
def how_to_multiply(num1):
    return lambda num2 : num1 * num2

double = how_to_multiply(2)
triple = how_to_multiply(3)

double(10)
triple(10)

30

#### If Statements within Lambdas

In [19]:
# with lambdas, if statements require an else

greater_than_10_add_2 = lambda num : num + 2 if num > 10 else num

# work around with elif
# greater_than_10_add_2 = lambda num : num + 2 if num > 10 else num (num - 2 if num == else num)

greater_than_10_add_2(11)

13

#### In-Class Exercise #1 <br>
<p>Write an anonymous function that cubes the arguments passed in and assign the anonymous function to a variable 'f'.</p>

In [31]:
f = lambda num : num**3 if num % 2 == 0 else num**2
    
print(f(10))
print(f(5))


1000
25


## Map <br>
<p>The map function allows you to iterate over an entire list while running a function on each item of the list. This is why the map function works well with lambda's, because it simplifies things and you write less lines of code.<br>The syntax for a map function is "map(function to be used, list to be used)"<br>However, you must be careful, as the map function returns a map object, not a list. To turn it into a list we use the list() type conversion.</p>

#### Syntax

In [35]:
 # map(function, iterable)
    
def square_number(num):
    return num ** 2

number_list = [1,2,3,4,5,6]

# Using list() function
list(map(square_number, number_list))

[1, 4, 9, 16, 25, 36]

#### Using Lambda's with Map

In [37]:
for num in map(lambda num : num ** 3, number_list):
    print(num, end=' ')
    
# same code with list comp

[num for num in map(lambda num : num ** 3, number_list)]

1 8 27 64 125 216 

[1, 8, 27, 64, 125, 216]

In [ ]:
attackers = [5,6,2,8]
defenders = [4,7,1,7]

survivors = list(map(lambda attacker, deffender : True if attacker < defender, attackers, defenders))
survivors.count(True) > survivorscount(False)

#### In-Class Exercise #2 <br>
<p>Use the map function to double each number and minus it by one in the list by using a lambda function</p>

In [38]:
number_list = [1,2,3,4,5,6,7,8,9,10]

[num for num in map(lambda num : (num * 2) - 1, number_list)]


nums = [2,3,5,2,3]
result =list(map(lambda num: num*2 - 1,nums))
print(result)

[1, 3, 5, 7, 9, 11, 13, 15, 17, 19]

## Filter() <br>
<p>Filter's are similar to the map function, where you're able to pass a function argument and a list argument and filter out something from the list based on the conditions passed. Similar to the map function, it returns a filter object, so you need to type convert it to a list()</p>

#### Syntax

In [40]:
# filter(function, iterable) returns filter object

def greater_than_nine(num):
    return num >= 10

num_list = [10,1,100,5,11]

filter(greater_than_nine, num_list)
list(filter(greater_than_nine, num_list)) # To print results 

[10, 100, 11]

#### Using Lambda's with Filter()

In [42]:
students = ['Hamed', 'Andy', 'William', 'Christopher', 'Will']

[student for student in filter(lambda student: student[0].lower() == 'w' , students)]

['William', 'Will']

#### In-Class Exercise #3 <br>
<p>Filter out all the numbers that are below the mean of the list.<br><b>Hint: Import the 'statistics' module</b></p>

In [59]:
from statistics import mean

num_list = [2,7,4.2,1.6,9,4.4,4.9]
average = mean(num_list)

print(num_list)
print([num for num in filter(lambda num : num <= average, num_list)])

[2, 7, 4.2, 1.6, 9, 4.4, 4.9]
[2, 4.2, 1.6, 4.4]


In [60]:
' '.join(filter(lambda word: word.count('!')!=1, '!!!Hi !!hi!!! !hi'.split())) # Codewars example

'!!!Hi !!hi!!!'

## Reduce() <br>
<p>Be very careful when using this function, as of Python 3 it's been moved to the 'functools' library and no longer is a built-in function.<br>The creator of Python himself, says to just use a for loop instead.</p>

#### Syntax

In [62]:
from functools import reduce

# reduce(function, iterable)

sum([1,2,3])

def add_two_numbers(num1, num2):
    return num1 + num2

reduce(add_two_numbers, [1,2,3,4,5])

15

#### Using Lambda's with Reduce()

In [63]:
reduce(lambda num1, num2 : num1 * num2, [1,2,3,4,5])

120

#### In-Class Exercise #4 <br>
<p>Use the reduce function to multiply the numbers in the list below together with a lambda function.</p>

In [64]:
reduce(lambda num1, num2 : num1 - num2, [1,2,3,4,5])

-13

## Recursion <br>
<p>Recursion means that a function is calling itself, so it contanstly executes until a base case is reached. It will then push the returning values back up the chain until the function is complete. A prime example of recursion is computing factorials... such that 5! (factorial) is 5*4*3*2*1 which equals 120.</p>

#### Implementing a Base Case

In [69]:
def add_num(num):
    if num == 1:
        return 1    
    return num + add_num(num -1)

add_num(5)

15

#### Writing a Factorial Function

In [68]:
def find_factorial(num):
    if num == 1:
        print(f'find_factorial({num}) = 1')
        return num
    print(f'find_factorial({num}) = {num} * find_factorial({num - 1})')
    return num * find_factorial(num - 1)

find_factorial(5)

find_factorial(5) = 5 * find_factorial(4)
find_factorial(4) = 4 * find_factorial(3)
find_factorial(3) = 3 * find_factorial(2)
find_factorial(2) = 2 * find_factorial(1)
find_factorial(1) = 1


120

#### In-Class Exercise #5 <br>
<p>Write a recursive function that subtracts all numbers to the argument given.</p>

In [79]:
def sub_num(num):
    if num == 1:
        return 1    
    return num - sub_num(num -1)

sub_num(10)

5

## Generators <br>
<p>Generators are a type of iterable, like lists or tuples. They do not allow indexing, but they can still be iterated through with for loops. They are created using functions and the yield statement.</p>

#### Yield Keyword <br>
<p>The yield keyword denotes a generator, it doesn't return so it won't leave the function and reset all variables in the function scope, instead it yields the number back to the caller.</p>

#### Infinite Generator

In [ ]:
# bad, never create infinite loops

#### In-Class Exercise #6 <br>
<p>Create a generator that takes a number argument and yields that number squared, then prints each number squared until zero is reached.</p>

# Exercises

### Exercise #1 <br>
<p>Filter out all of the empty strings from the list below</p>

`Output: ['Argentina', 'San Diego', 'Boston', 'New York']`

In [138]:
places = [" ","Argentina", " ", "San Diego","","  ","","Boston","New York"]

[place for place in filter(lambda empty_string: empty_string != ''and empty_string != ' ' and empty_string != '  ', places)]

['Argentina', 'San Diego', 'Boston', 'New York']

### Exercise #2 <br>
<p>Write an anonymous function that sorts this list by the last name...<br><b>Hint: Use the ".sort()" method and access the key"</b></p>

`Output: ['Victor aNisimov', 'Gary A.J. Bernstein', 'Joel Carter', 'Andrew P. Garfield', 'David hassELHOFF']`

In [10]:
author = ["Joel Carter", "Victor aNisimov", "Andrew P. Garfield","David hassELHOFF","Gary A.J. Bernstein"]
s_authors = {}


author.sort(reverse=False, key=lambda name : name.split()[-1].lower()) #descending order with reverse

print(author)

#lambda 

['Victor aNisimov', 'Gary A.J. Bernstein', 'Joel Carter', 'Andrew P. Garfield', 'David hassELHOFF']


### Exercise #3 <br>
<p>Convert the list below from Celsius to Farhenheit, using the map function with a lambda...</p>

`Output: [('Nashua', 89.6), ('Boston', 53.6), ('Los Angelos', 111.2), ('Miami', 84.2)]
`

In [4]:
# F = (9/5)*C + 32
places = [('Nashua',32),("Boston",12),("Los Angelos",44),("Miami",29)]


places = [('Nashua',32),("Boston",12),("Los Angelos",44),("Miami",29)]

[place for place in map(lambda place : (place[0], (9/5)*place[1]+32), places)] #have to identify index number to start

[('Nashua', 89.6), ('Boston', 53.6), ('Los Angelos', 111.2), ('Miami', 84.2)]

### Exercise #4 <br>
<p>Write a recursion function to perform the fibonacci sequence up to the number passed in.</p>

`Output for fib(5) => 
Iteration 0: 1
Iteration 1: 1
Iteration 2: 2
Iteration 3: 3
Iteration 4: 5
Iteration 5: 8`

In [8]:
def fib_sequence(num):
    if num <= 1:
        return num
    return fib_sequence(num - 2) + fib_sequence(num - 1)

def fib(n):
    for i in range(n + 1):
        print(f"Iteration {i}: {fib_sequence(i + 1)}", end = ' ')

fib(5)

Iteration 0: 1 Iteration 1: 1 Iteration 2: 2 Iteration 3: 3 Iteration 4: 5 Iteration 5: 8 